samples/boards/nrf52/mesh/onoff-app/src/main.c

```diff
@@ -0,0 +1,694 @@
+				onoff_state->current);
+		onoff_state->previous = onoff_state->current;
+		bt_mesh_model_msg_init(msg,
+				BT_MESH_MODEL_OP_GEN_ONOFF_STATUS);
+		net_buf_simple_add_u8(msg, onoff_state->current);
+		err = bt_mesh_model_publish(model);
+		if (err) {
+			printk("%s bt_mesh_model_publish: err: %d\n",
+					__func__, err);
+		}
+	}
+}
+
+static void gen_onoff_set(struct bt_mesh_model *model,
+			  struct bt_mesh_msg_ctx *ctx,
+			  struct net_buf_simple *buf)
+{
+	printk("%s\n", __func__);
+
+	gen_onoff_set_unack(model, ctx, buf);
+	gen_onoff_get(model, ctx, buf);
+}
+
+static void gen_onoff_status(struct bt_mesh_model *model,
+			     struct bt_mesh_msg_ctx *ctx,
+			     struct net_buf_simple *buf)
+{
+	u8_t	state;
+
+	state = net_buf_simple_pull_u8(buf);
+
+	printk("Node 0x%04x received OnOff status from 0x%04x with state 0x%02x\n",
+			model->elem->addr, ctx->addr, state);
+}
+
+
+
+static int output_number(bt_mesh_output_action_t action, uint32_t number)
+{
+	printk("OOB Number: %u\n", number);
+
+	return 0;
+}
+
+static int output_string(const char *str)
+{
+	printk("OOB String: %s\n", str);
+	return 0;
+}
+
+static void prov_complete(u16_t net_idx, u16_t addr)
+{
+	printk("provisioning complete for net_idx: 0x%04x addr: 0x%04x\n",
+			net_idx, addr);
+	primary_addr = addr;
+	primary_net_idx = net_idx;
+}
+
+static void prov_reset(void)
+{
+	bt_mesh_prov_enable(BT_MESH_PROV_ADV | BT_MESH_PROV_GATT);
+}
+
+static u8_t dev_uuid[16] = { 0xdd, 0xdd };
+
+/*
+ * Button Pressed Interrupt Handler
+ */
+
+#define BUTTON_DEBOUNCE_DELAY_MS 250
+
+static void button_pressed(struct device *dev, struct gpio_callback *cb,
+		 uint32_t pins)
+{
+	time = k_uptime_get_32();
+
+	/* debounce the switch */
+	if (time < last_time + BUTTON_DEBOUNCE_DELAY_MS) {
+		return;
+	}
+
+	last_time = time;
+	sw.pin = pins;
+	k_work_submit(&sw.button_work);
+}
+
+/*
+ * Map GPIO pins to button number
+ */
+
+static uint8_t pin_to_sw(uint32_t pin)
+{
+
+	switch (pin) {
+	case BIT(SW0_GPIO_PIN): return 0;
+	case BIT(SW1_GPIO_PIN): return 1;
+	case BIT(SW2_GPIO_PIN): return 2;
+	case BIT(SW3_GPIO_PIN): return 3;
+	}
+	printk("no match for pin\n");
+	return 0;
+}
+
+/*
+ * Button Pressed Worker Task
+ */
+
+static void button_pressed_worker(struct k_work *work)
+{
+	struct bt_mesh_model *mod_cli, *mod_srv;
+	struct bt_mesh_model_pub *pub_cli, *pub_srv;
+	struct sw *sw = CONTAINER_OF(work, struct sw, button_work);
+	int err;
+	u8_t sw_idx = pin_to_sw(sw->pin);
+
+	mod_cli = mod_cli_sw[sw_idx];
+	pub_cli = mod_cli->pub;
+
+	mod_srv = mod_srv_sw[sw_idx];
+	pub_srv = mod_srv->pub;
+
+	/* Invert onoff state */
+	onoff_state[sw_idx].current = onoff_state[sw_idx].current ? 0 : 1;
+
+	printk("%s pin: 0x%08x sw: 0x%02x onoff: 0x%02x\n",
+			__func__, sw->pin, sw_idx, onoff_state[sw_idx].current);
+
+	/* pin = 0 turns on LED's on nrf52840-pca10056 board */
+	gpio_pin_write(onoff_state[sw_idx].led_device, onoff_state[sw_idx].led_gpio_pin,
+			onoff_state[sw_idx].current ? 0 : 1);
```

Comment:
It's possible I've understood the code wrong, but is the reason you're having to call gpio_pin_write() here the fact that you're ignoring locally originated messages in the gen_onoff_set_unack() handler? Wouldn't it be simpler to handle locally originated messages there just like any other message, i.e. the toggling of the LED would only happen from one place then.

Reply:
I made the changes you suggested and it greatly simplified things.

While working on this, I realized that having the button just change the state of the led doesn't make sense. The button knows it's last state, but nothing about other leds in a group. I'm thinking of changing the button's behavior to recognize 1 button push in a second as on and 2 as off. Can you can think of a better way?

Alternatively, I could have button 1 & 2 send on & off respectively. The same for button 3 & 4. With that approach, only 2 elements would have button clients, all 4 would have led servers. The latter might better reflect the real world.

Reply:
I changed the button press logic to use 1 press for on and > 1 for off within a 1 second interval. It's a little clumsy, but I wanted to keep the 4 client models. I changed the readme to reflect this.

When unprovisioned, I use the server function to change the led state. Much simpler.

I've left most of the instrumentation as an aid in seeing what's going on internally.

Luiz expects to push the meshctl changes today.

Thanks a lot for helping me with this.